src/main/java/eu/europeana/metis/sandbox/service/workflow/EnrichmentServiceImpl.java

```diff
@@ -3,41 +3,64 @@
+import java.util.Set;
+import java.util.Objects;
+
 import org.springframework.stereotype.Service;
 
 @Service
 class EnrichmentServiceImpl implements EnrichmentService {
 
-  private final EnrichmentWorker enrichmentWorker;
-
-  public EnrichmentServiceImpl(EnrichmentWorker enrichmentWorker) {
-    this.enrichmentWorker = enrichmentWorker;
-  }
-
-  @Override
-  public RecordInfo enrich(Record record) {
-    requireNonNull(record, "Record must not be null");
-
-    List<RecordError> recordErrors = new LinkedList<>();
-    byte[] result;
-    try {
-      result = enrichmentWorker.process(record.getContentInputStream());
-    } catch (EnrichmentException|DereferenceException|SerializationException e) {
-      result = record.getContent();
-      recordErrors.add(new RecordError(new RecordProcessingException(record.getProviderId(), e)));
-    } catch (Exception e) {
-      throw new RecordProcessingException(record.getProviderId(), e);
+    private final EnrichmentWorker enrichmentWorker;
+
+    public EnrichmentServiceImpl(EnrichmentWorker enrichmentWorker) {
+        this.enrichmentWorker = enrichmentWorker;
+    }
+
+    @Override
+    public RecordInfo enrich(Record record) {
+        requireNonNull(record, "Record must not be null");
+
+        List<RecordError> recordErrors = new LinkedList<>();
+        ProcessedResult<byte[]> result;
+        try {
+            result = enrichmentWorker.process(record.getContentInputStream());
+            Set<Report> reports = result.getReport();
+
+            if (result.getRecordStatus().equals(ProcessedResult.RecordStatus.STOP)) {
+                Optional<Report> report = reports.stream().filter(rep -> Objects.equals(rep.getMessageType(), Type.ERROR)).findFirst();
+                throw new RecordProcessingException(record.getProviderId(), handleRecordStopException(report));
+            }
+
+            reports.stream().filter(report -> report.getMessageType().equals(Type.WARN))
+                    .forEach(report -> recordErrors.add(new RecordError(new RecordProcessingException(record.getProviderId(),
+                            new ServiceException(handleErrorMessage(report), null)))));
+
+        } catch (Exception e) {
+            throw new RecordProcessingException(record.getProviderId(), e);
+        }
+
+        return new RecordInfo(Record.from(record, result.getProcessedRecord()), recordErrors);
     }
 
-    return new RecordInfo(Record.from(record, result), recordErrors);
-  }
+    private String handleErrorMessage(Report report) {
+        return String.format("%s Value: %s", report.getMessage() ,report.getValue());
+    }
+
+    private Exception handleRecordStopException(Optional<Report> report) {
+        return report.isPresent() ? new ServiceException(handleErrorMessage(report.get()),
+                new Exception("StackTrace: " + report.get().getStackTrace())) :
+                new ServiceException("Something went wrong when requesting report.", null);
```

Comment:
My IDE suggests replacing this with 
```
return report
    .map(value -> new ServiceException(createErrorMessage(value), new Exception("StackTrace: " + value.getStackTrace())))
    .orElseGet(() -> new ServiceException("Something went wrong when requesting report.", null));
```
This to keep benefiting from the functional nature of optionals.
What do you think?

Also: do we need that innermost exception (with the stacktrace as argument)? or should we just pass null, just like we do with the `ServiceException` in the `Type.WARN` case?

Reply:
My IDE was also suggesting it, but decided to keep what I had originally so it would be easier to understand what is happening. But I've changed to method because of previous comments, so this looks different now.

That's a good point, I was thinking that perhaps it would be useful to save the Stack trace somewhere for future debugging, but I can change it null